x-pack/plugin/searchable-snapshots/src/main/java/org/elasticsearch/index/store/SearchableSnapshotDirectory.java

```diff
@@ -393,33 +403,64 @@ protected void doRun() throws Exception {
+                                warmer.onFailure(new IOException("Shard cache prewarming cancelled"));
                             }
+                        } catch (InterruptedException ie) {
+                            Thread.currentThread().interrupt();
+                        }
+                    }
+                } finally {
+                    if (enqueued == false) {
+                        IOUtils.closeWhileHandlingException(input);
+                    }
+                }
+            }
 
-                            @Override
-                            public void onFailure(Exception e) {
-                                logger.trace(
-                                    () -> new ParameterizedMessage(
-                                        "failed to warm cache for [{}] part [{}/{}]",
-                                        fileName,
-                                        part,
-                                        numberOfParts
-                                    ),
-                                    e
-                                );
-                            }
+            // Start as many workers as fit into the searchable snapshot pool at once at the most
+            final int workers = Math.min(threadPool.info(SEARCHABLE_SNAPSHOTS_THREAD_POOL_NAME).getMax(), queue.size());
+            for (int i = 0; i < workers; ++i) {
+                executeNextWarmer(executor, queue);
+            }
+        }
+    }
 
-                            @Override
-                            public void onAfter() {
-                                if (countDown.countDown()) {
-                                    IOUtils.closeWhileHandlingException(input);
-                                }
-                            }
-                        });
+    private void executeNextWarmer(final Executor executor, final BlockingQueue<ActionRunnable<Void>> queue) {
+        try {
+            final ActionRunnable<?> warmer = queue.poll(0L, TimeUnit.MILLISECONDS);
+            if (warmer != null) {
+                executor.execute(new AbstractRunnable() {
+                    @Override
+                    protected void doRun() {
+                        warmer.run();
+                        executeNextWarmer(executor, queue);
                     }
-                } catch (IOException e) {
-                    logger.trace(() -> new ParameterizedMessage("failed to warm cache for [{}]", fileName), e);
-                }
+
+                    @Override
+                    public void onRejection(Exception e) {
+                        warmer.onRejection(e);
+                    }
+
+                    @Override
+                    public void onFailure(Exception e) {
+                        logger.warn("failed to schedule next file part to prewarm in cache", e);
+                    }
+                });
             }
+        } catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
```

Comment:
This just swallows the exception?

Reply:
I don't think we need to do more than logging here so I pushed b6b90c4e6c630972a8888dbadbc8222cf2ecc9bc